compiler/luci/service/src/CircleShapeSignatureInferenceHelper.cpp

```diff
@@ -16,12 +16,133 @@
 
 #include "luci/Service/CircleShapeSignatureInferenceHelper.h"
 
+#include <loco.h>
+
+#include <luci/Log.h>
+
+#include <oops/InternalExn.h>
+
+namespace
+{
+
+luci::ShapeSignature clean_signature(const luci::ShapeSignature &signature)
+{
+  // If shape signature has at least one -1, it is not static.
+  for (uint32_t i = 0; i < signature.rank(); ++i)
+    if (signature.dim(i) == -1)
+      return signature;
+
+  // If all dimensions are static, return empty shape signature.
+  return luci::ShapeSignature();
+}
+
+} // namespace
+
 namespace luci
 {
 
 namespace ssinf
 {
 
+ShapeSignature reduced_signature(const loco::Node *node, const loco::Node *indices, bool keep_dims)
+{
+  ShapeSignature output_signature;
+
+  auto circle_node = loco::must_cast<const luci::CircleNode *>(node);
+  auto circle_indices = loco::must_cast<const luci::CircleNode *>(indices);
+  auto input_signature = circle_node->shape_signature();
+
+  // If shape signature is empty, output is static.
+  if (input_signature.rank() == 0)
+    return output_signature;
+
+  auto reduction_indices = dynamic_cast<const luci::CircleConst *>(circle_indices);
+
+  // When reduction_indices is not constant
+  if (reduction_indices == nullptr)
+  {
+    if (keep_dims)
+    {
+      // If keep_dims is true, rank is not changed.
+      output_signature.rank(input_signature.rank());
+      for (uint32_t i = 0; i < output_signature.rank(); ++i)
+        output_signature.dim(i) = -1;
+    }
```

Comment:
Maybe this line seems redundant because of below line.
```cxx
  if (keep_dims)
  {
    output_signature.rank(input_signature.rank());
    for (uint32_t i = 0; i < input_signature.rank(); ++i)
      output_signature.dim(i) = input_signature.dim(i);
    for (uint32_t i = 0; i < reduction_values.size(); ++i)
      output_signature.dim(reduction_values.at(i)) = 1;
  }
```

Reply:
These two codes looks different...? And this block ends at line 79

Reply:
:cold_sweat: